### PR TITLE
SQL: Implement DATE_PARSE function for parsing strings into DATE values

### LINKED ISSUE #54962: SQL: Implement DATE_PARSE
Implement DATE_PARSE(<date_str>, <pattern_str>) function
which allows to parse a date string according to the specified
pattern into a date object. The patterns allowed are those of
`java.time.format.DateTimeFormatter`.

Follows: #54960

### PR DESCRIPTION
**Pull request by Patrick0308**
Implement DATE_PARSE(<time_str>, <pattern_str>) function
which allows to parse a time string according to the specified
pattern into a time object. The patterns allowed are those of
java.time.format.DateTimeFormatter.

Closes #54962

### CODE REVIEW

**Comment by Patrick0308** [comment]
@matriv I have finished this work. Please review it.

**Review by matriv — COMMENTED** [review]
Thank you for picking this up. I left some comments.
Please take care of the code formatting, and the copy-paste errors.

File: x-pack/plugin/sql/qa/server/src/main/resources/date.csv-spec [hunk 1]
```diff
@@ -164,3 +164,105 @@ SELECT MAX(salary) FROM test_emp GROUP BY TODAY();
...
+
+   date1       
+------------
+2020-04-07 
+;
+
+
+selectDateParseWithField
+schema::birth_date:ts|dp_birth_date:date
+SELECT birth_date, DATE_PARSE(DATETIME_FORMAT(birth_date, 'MM/dd/ HH uuuu'), concat(gender, 'M/dd/ HH uuuu')) AS dp_birth_date
```
[the remark this answers is not in the document]
Comment (Patrick0308) [reply x-pack/plugin/sql/qa/server/src/main/resources/date.csv-spec:180]: It's work. I found if the parsing pattern contains time units and date units (e.g. 'MM/dd/ HH mm ss  uuuu', 'MM/dd/ HH  uuuu', etc.) an error will not returned,

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeParseProcessor.java [hunk 2]
```diff
@@ -31,7 +32,8 @@
 
     public enum Parser {
         DATE_TIME("datetime", ZonedDateTime::from, LocalDateTime::from), 
-        TIME("time", OffsetTime::from, LocalTime::from);
+        TIME("time", OffsetTime::from, LocalTime::from),
+        DATE("date", LocalDate::from, (TemporalAccessor ta) -> {throw new RuntimeException();});
```
[the remark this answers is not in the document]
Comment (Patrick0308) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeParseProcessor.java:36]: `DateTimeFormatter.parseBest` needs two queries at least. May be adding `(TemporalAccessor ta) -> {throw new RuntimeException();}` in constructor is better?

File: docs/reference/sql/functions/date-time.asciidoc [hunk 3]
```diff
@@ -404,6 +404,49 @@ include-tagged::{sql-specs}/docs/docs.csv-spec[dateDiffDateTimeMinutes]
...
+
+*Output*: date
+
+*Description*: Returns a date by parsing the 1st argument using the format specified in the 2nd argument. The parsing
+format pattern used is the one from
+https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/time/format/DateTimeFormatter.html[`java.time.format.DateTimeFormatter`].
+If any of the two arguments is `null` or an empty string `null` is returned.
+
+[NOTE]
+If the parsing pattern not contains date units completely (e.g. 'HH:mm:ss', 'dd-MM HH:mm:ss', etc.) an error is returned
```
Comment (matriv) [inline docs/reference/sql/functions/date-time.asciidoc:431]:
```suggestion
If the parsing pattern does not contain all valid date units (e.g. 'HH:mm:ss', 'dd-MM HH:mm:ss', etc.) an error is returned
```

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeParseProcessor.java [hunk 4]
```diff
@@ -31,16 +32,21 @@
 
     public enum Parser {
         DATE_TIME("datetime", ZonedDateTime::from, LocalDateTime::from),
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeParseProcessor.java:34]:
```suggestion
        DATE_TIME(DataTypes.DATETIME, ZonedDateTime::from, LocalDateTime::from),
        TIME(SqlDataTypes.TIME, OffsetTime::from, LocalTime::from),
        DATE(SqlDataTypes.DATE, LocalDate::from, (TemporalAccessor ta) -> { throw new DateTimeException("InvalidDate");});
        
        private final BiFunction<String, String, TemporalAccessor> parser;
        
        private final String parseType;

        Parser(DataType parseType,  TemporalQuery<?>... queries) {
            this.parseType = parseType.typeName();
            this.parser = (timestampStr, pattern) -> DateTimeFormatter.ofPattern(pattern, Locale.ROOT).parseBest(timestampStr, queries);
        }
```

**Review by matriv — COMMENTED** [review]
Left some comments.
Let's keep accepting time parts together with date parts as the DateTimeFormatter generates the correct date.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeParseProcessor.java [hunk 5]
```diff
@@ -30,15 +34,17 @@
...
+        DATE(SqlDataTypes.DATE, LocalDate::from, (TemporalAccessor ta) -> {throw new DateTimeException("InvalidDate");});
         
         private final BiFunction<String, String, TemporalAccessor> parser;
         
-        private final String parseType;
+        private final DataType parseType;
 
-        Parser(String parseType,  TemporalQuery<?>... queries) {
+        Parser(DataType parseType, TemporalQuery<?>... queries) {
             this.parseType = parseType;
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeParseProcessor.java:46]:
```suggestion
            this.parseType = parseType.typeName();
```

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeParseProcessor.java [hunk 6]
```diff
@@ -30,15 +34,17 @@
...
-        DATE_TIME("datetime", ZonedDateTime::from, LocalDateTime::from), 
-        TIME("time", OffsetTime::from, LocalTime::from);
+        DATE_TIME(DataTypes.DATETIME, ZonedDateTime::from, LocalDateTime::from), 
+        TIME(SqlDataTypes.TIME, OffsetTime::from, LocalTime::from),
+        DATE(SqlDataTypes.DATE, LocalDate::from, (TemporalAccessor ta) -> {throw new DateTimeException("InvalidDate");});
         
         private final BiFunction<String, String, TemporalAccessor> parser;
         
-        private final String parseType;
+        private final DataType parseType;
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeParseProcessor.java:43]:
```suggestion
        private final String parseType;
```

**Review by matriv — COMMENTED** [review]
Left some more comments.

**Comment by matriv** [comment]
@elasticmachine test this please

File: x-pack/plugin/sql/qa/server/src/main/resources/docs/docs.csv-spec [hunk 7]
```diff
@@ -2864,6 +2865,31 @@ schema::time:time
...
+
+   date
+-----------
+2020-04-07
+// end::dateParse1
+;
+
+dateParse2-Ignore
+schema::date:date
+// tag::timeParse3
```
Comment (matriv) [inline x-pack/plugin/sql/qa/server/src/main/resources/docs/docs.csv-spec:2881]:
```suggestion
// tag::dateParse2
```

**Review by matriv — COMMENTED** [review]
Left a couple more comments.

**Comment by matriv** [comment]
@elasticmachine test this please

**Comment by Patrick0308** [comment]
@elasticmachine update this branch

**Comment by Patrick0308** [comment]
@elasticmachine update branch

**Comment by matriv** [comment]
@elasticmachine test this please

**Comment by matriv** [comment]
@elasticmachine update branch

**Comment by matriv** [comment]
@elasticmachine test this please

**Review by matriv — APPROVED** [review]
LGTM

File: docs/reference/sql/functions/date-time.asciidoc [hunk 8]
```diff
@@ -404,6 +404,50 @@ include-tagged::{sql-specs}/docs/docs.csv-spec[dateDiffDateTimeMinutes]
...
+
+<1> date expression as a string
+<2> parsing pattern
+
+*Output*: date
+
+*Description*: Returns a date by parsing the 1st argument using the format specified in the 2nd argument. The parsing
+format pattern used is the one from
+https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/time/format/DateTimeFormatter.html[`java.time.format.DateTimeFormatter`].
+If any of the two arguments is `null` or an empty string `null` is returned.
```
Comment (bpintea) [inline docs/reference/sql/functions/date-time.asciidoc:428]:
```suggestion
If any of the two arguments is `null` or an empty string, then `null` is returned.
```

**Review by bpintea — APPROVED** [review]
LGTM

**Comment by matriv** [comment]
@elasticmachine test this please

**Comment by matriv** [comment]
@elasticmachine update branch

**Comment by matriv** [comment]
@elasticmachine test this please

**Pull request merged**